archive-data/archive.py

```diff
@@ -0,0 +1,204 @@
+
+    logging.basicConfig(level=logging.INFO,
+                        format='%(asctime)s:%(levelname)s:%(message)s')
+
+
+def get_db_connection(config: dict[str, str]) -> connection:
+    """Returns a connection to a database."""
+
+    return connect(
+        host=config["DB_HOST"],
+        user=config["DB_USER"],
+        password=config["DB_PASS"],
+        dbname=config["DB_NAME"],
+        port=config["DB_PORT"],
+        cursor_factory=RealDictCursor
+    )
+
+
+def close_connection(conn: connection) -> None:
+    """Closes the connection to the database."""
+
+    conn.close()
+
+
+def fetch_old_data(conn: connection, query: str) -> pd.DataFrame:
+    """Fetches historical data based on the query.
+    Returns the fetched data as a DataFrame."""
+
+    try:
+
+        with conn.cursor() as cur:
+            cur.execute(query)
+            rows = cur.fetchall()
+        return pd.DataFrame(rows)
+
+    except Error as err:
+
+        conn.rollback()
+        logging.error("Failed to fetch data: %s", err)
+        if err.pgcode:
+            logging.error("Error code: %s", err.pgcode)
+
+
+def get_stations_performance(conn: connection, queries: list[str]) -> pd.DataFrame:
+    """Fetches the historical data for the stations based on the queries."""
+
+    performance_data = []  # list for adding the dataframes
+
+    for query in queries:
+        result = fetch_old_data(conn, query)
+        if not result.empty:
+            performance_data.append(result)
+
+    if not performance_data:
+        raise DataFetchError("No data fetched from the database.")
+
+    # combines the dataframes
+    return reduce(lambda left, right: pd.merge(left, right, on=['day', 'station_id'], how='outer'), performance_data)
+
+
+def get_operators_performance(conn: connection, queries: list[str]) -> pd.DataFrame:
+    """Fetches the historical data for the operators based on the queries."""
+
+    performance_data = []  # list for adding the dataframes
+
+    for query in queries:
+        result = fetch_old_data(conn, query)
+        if not result.empty:
+            performance_data.append(result)
+
+    if not performance_data:
+        raise DataFetchError("No data fetched from the database.")
+
+    # combines the dataframes
+    return reduce(lambda left, right: pd.merge(left, right, on=['day', 'operator_id'], how='outer'), performance_data)
+
+
+def clean_data(data: pd.DataFrame) -> pd.DataFrame:
+    """Cleans the data by converting the columns to the
+    correct data types and filling in the missing values."""
+
+    data['day'] = pd.to_datetime(data['day']).dt.date
+    data['delay_1m_count'] = data['delay_1m_count'].fillna(0).astype(int)
+    data['delay_5m_count'] = data['delay_5m_count'].fillna(0).astype(int)
+    data['avg_delay_min'] = data['avg_delay_min'].fillna(0).astype(int)
+    data['arrival_count'] = data['arrival_count'].fillna(0).astype(int)
+    data['cancellation_count'] = data['cancellation_count'].fillna(0).astype(int)
```

Comment:
Repetitive code - this could be put into a function maybe? The structure for each of these lines is the same:`data[col] = data[col].fillna(0).astype(int)`